perf: make transaction summary not wait for top-level query

alxu-sentry opened the pull request:
LoadingIndicator was added in #20956, probably because we need p95 to get the
transactions list. However, this unnecessarily holds up all the other API
requests. Optimize the transaction summary page to start those requests first
and wait for the p95 to get the transactions list.

While we're here, also migrate some things from em dashes to Placeholders to
improve consistency on this page.

alxu-sentry committed:
perf: make transaction summary not wait for top-level query

LoadingIndicator was added in #20956, probably because we need p95 to get the
transactions list. However, this unnecessarily holds up all the other API
requests. Optimize the transaction summary page to start those requests first
and wait for the p95 to get the transactions list.

While we're here, also migrate some things from em dashes to Placeholders to
improve consistency on this page.

silent1mezzo commented:
@alxu-sentry In the future can we keep PRs to related changes (e.g. Placeholders can be done in one PR, performance enhancements in another). This makes it easier to review.

dashed commented on src/sentry/static/sentry/app/views/performance/transactionSummary/index.tsx:
```diff
@@ -196,13 +182,10 @@ class TransactionSummary extends React.Component<Props, State> {
                 location={location}
               >
-                {({tableData, isLoading}) => {
-                  if (isLoading) {
```
@alxu-sentry Is this where the block was occurring?

alxu-sentry replied:
yes. if we don't render any subcomponents then they won't issue their `<DiscoverQuery>`s

dashed commented on src/sentry/static/sentry/app/views/performance/transactionSummary/content.tsx:
```diff
@@ -204,22 +202,27 @@ class SummaryContent extends React.Component<Props, State> {
+                handleBaselineClick={this.handleViewDetailsClick}
+                handleCellAction={this.handleCellAction}
+                handleOpenInDiscoverClick={this.handleDiscoverViewClick}
+                {...getTransactionsListSort(location, {
```
@alxu-sentry It'll be better to relocate `getTransactionsListSort()` function into the same file as the `<TransactionsList/>` component.

alxu-sentry replied:
Are you sure? Only Releases and Transaction Summary use TransactionsList at the moment, in pretty different ways. I think it'd be better to keep it here until we have more users and a clearer view of what a unified API would look like.

dashed commented on src/sentry/static/sentry/app/views/performance/transactionSummary/index.tsx:
```diff
@@ -196,13 +182,10 @@ class TransactionSummary extends React.Component<Props, State> {
                   const totals = (tableData && tableData.data.length
                     ? tableData.data[0]
-                    : emptyValues) as TotalValues;
+                    : null) as TotalValues;
```
```suggestion
                    : null) as TotalValues | null;
```

dashed commented on src/sentry/static/sentry/app/views/performance/transactionSummary/content.tsx:
```diff
@@ -204,22 +202,27 @@ class SummaryContent extends React.Component<Props, State> {
-              handleOpenInDiscoverClick={this.handleDiscoverViewClick}
-            />
+            {totalValues === null ? (
+              <Placeholder height="352px" />
```
@alxu-sentry This placeholder is far too large. It's preferable to delegate the placeholder to the `<TransactionsList />` component. It already has a decent loading state from the discover query component inside. That would provide better loading aesthetics.

alxu-sentry replied:
addressed this issue in https://github.com/getsentry/sentry/pull/24298/commits/86f9e82ab5c00351813e64f719859de5cba84ae6

alxu-sentry committed:
Update src/sentry/static/sentry/app/views/performance/transactionSummary/index.tsx

Co-authored-by: Alberto Leal <mail4alberto@gmail.com>

alxu-sentry committed:
use TransactionsList loading indicator instead of a placeholder

add a forceLoading prop to TransactionsList and use that to bypass the query.
the downside of this method is that even if the user selects a different sort
option that doesn't require the top-level query, it will still wait. this
shouldn't be a huge issue on reasonable speed internet though.

Zylphrex commented on src/sentry/static/sentry/app/views/performance/transactionSummary/content.tsx:
```diff
@@ -202,27 +201,24 @@ class SummaryContent extends React.Component<Props, State> {
+              handleCellAction={this.handleCellAction}
+              handleOpenInDiscoverClick={this.handleDiscoverViewClick}
+              {...getTransactionsListSort(location, {
+                p95: totalValues ? totalValues.p95 : 0,
```
A little trick with optional chaining and the nullish coalescing operator. You can even chain a bunch of these together like `a?.b?.c?.d`.

```suggestion
                p95: totalValues?.p95 ?? 0,
```

alxu-sentry replied:
hm, i knew about `?.` but forgot about `??`. actually now that i think about it, `totalValues?.p95 || 0` would actually correct here but unnecessarily confusing

Zylphrex commented on tests/js/spec/components/discover/transactionsList.spec.jsx:
```diff
@@ -373,4 +373,22 @@ describe('TransactionsList', function () {
+          forceLoading
+        />
+      );
+      expect(wrapper.find('LoadingIndicator')).toHaveLength(1);
```
Can you also include a prop change in this test to toggle `forceLoading` to false to assert that the list renders as expected?

alxu-sentry replied:
addressed in 675ae7b3b653cd2e2af9164cc3e44dcb79f712c6, the intent was to check that the api call is not made, but i don't think it worked properly anyways.

alxu-sentry committed:
Update src/sentry/static/sentry/app/views/performance/transactionSummary/content.tsx

Co-authored-by: Tony <txiao@sentry.io>

alxu-sentry committed:
update test

1. check that api is not called (was previous test's intention but was
   implemented wrong)
2. check that turning off forceLoading allows rendering

dashed reviewed (approved):
👏
